### ISSUE TITLE
Memory consumption

### ISSUE THREAD

**Issue by Paulie14** [flow123d/flow123d#875]
Motivated by the problem at: 
`bacula.nti.tul.cz/intec/lide/jirina.kralovcova/00_share/2017_Vypocet_flow3_06`
which can be solved with 2.1.2 and not in 3.0.0 due to larger memory consumption.

Smaller test (33k elements) was run:
`bacula.nti.tul.cz/intec/lide/pavel.exner/00_share/flow_problems/2017_memory_3_0_0`
which showed the following:
- over 10 % increase in Petsc allocation `MatSeqAIJSetPreallocation_SeqAIJ` in `LinSys_PETSC::preallocate_matrix() (linsys_PETSC.cc:237)`
- computation in schur complement takes additional memory in 3.0.0_dev `MatAXPY_SeqAIJ` in `SchurComplement::form_schur() (schur.cc:218)`
- Balance takes quite a lot of memory and allocation is called 3 times `Balance::lazy_initialize() (balance.cc:282)`
- we do not know why allocation `MatILUFactorSymbolic_SeqAIJ` in `LinSys_PETSC::solve() (linsys_PETSC.cc:381)` is called 4 times
- in 2.1.2 max peak is 284.3 MB, in 3.0.0_dev max peak is 320.5 MB

See the results at bacula link, use massif visualizer:
```
add-apt-repository ppa:kubuntu-ppa/backports
apt-get update && apt-get install massif-visualizer
```
Increase of the number of detailed snapshots in massif might be necessary.

**Comment by Paulie14** [comment]
I also tried MatInfo on schur0 matrix:
```
        MatInfo info;
        MatGetInfo(*schur0->get_matrix(),MAT_GLOBAL_SUM,&info);
        DBGVAR(info.nz_allocated);    // 2.67694e+06
        DBGVAR(info.nz_used);         // 2.61379e+06
        DBGVAR(info.nz_unneeded);     // 63150
```
The unneeded nz are 2,36%, so the larger allocation itself might not be the main issue.

**Comment by jstebel** [comment]
Try to reduce allocated memory by setting ```MatSetOption(m, MAT_IGNORE_ZERO_ENTRIES, PETSC_TRUE)``` where applicable.

**Comment by Paulie14** [comment]
`MatSetOption(m, MAT_IGNORE_ZERO_ENTRIES, PETSC_TRUE)` works great, from 320 MB peak I got to 270 MB peak.
HOWEVER I have problem with unsteady or nonlinear cases - e.g. seepage problem in test 11. The matrix is preallocated, then assembled and solved - at first step with dirichlet BC, then at second step, bc changes to neumann and there is not enough nz values allocated on the row..

**Comment by Paulie14** [comment]
Improved by setting `MatSetOption(m, MAT_IGNORE_ZERO_ENTRIES, PETSC_TRUE)`
and using additional 'almost zeros' in assembly to keep matrix structure.
Done in PE_300_leaks by commit 149a66b4a64736601186b23e253e52a15db47d7f.

**Comment by jbrezmorf** [comment]
resolved by  #884
